site/content/contribute/more-info/server/schema-migration-guide.md

```diff
@@ -22,15 +39,293 @@ A file naming convention is used to determine the order in which the migrations
+
+- **Data migrations**:
+
+```sql
+UPDATE channelmembers SET MentionCountRoot = ..
+UPDATE Channels SET TotalMsgCountRoot = ..
+UPDATE ChannelMembers CM SET MsgCountRoot ..
+```
+
+In these cases, rather than operating on the entire table, we need to operate on batches at a time. See the example later on how to achieve that.
+
+- **Changing NULL columns to NON-NULL**
+
+```sql
+UPDATE Channels SET LastRootPostAt=0 WHERE LastRootPostAt IS NULL;
+UPDATE OAuthApps SET MattermostAppID = '' WHERE MattermostAppID IS NULL;
+```
+
+This is possible to handle from the code itself using a `COALESCE` function. It makes the code more complicated, but it’s a cost we have to pay to reduce migration overhead.
+
+- **Denormalization optimizations**:
+
+```sql
+UPDATE threads SET threaddeleteat = posts.deleteat FROM posts WHERE threads.threaddeleteat IS NULL AND posts.id = threads.postid;
+UPDATE reactions SET channelid = COALESCE((select channelid from posts where posts.id = reactions.postid), '') WHERE channelid='';
+UPDATE threads SET threadteamid = channels.teamid FROM channels WHERE threads.threadteamid IS NULL AND channels.id = threads.channelid;
+UPDATE fileinfo SET channelid = posts.channelid FROM posts WHERE fileinfo.channelid IS NULL AND fileinfo.postid = posts.id;
+```
+
+We can take the same approach as in data migrations.
+
+9. DELETE
+
+So far, there have been only a handful of DELETE statements in schema migrations. And mostly they have been for security issues. The general recommendation is to avoid running a full-blown DELETE statement that operates on the entire table, but rather operate on batches so as to avoid taking a lock on the entire table. This could either be done in a job since there is no new code waiting for this to be executed. (See above)
+
+## Examples
+
+### How do I change a column type if I MUST
+
+Follow this long-winded procedure:
+- Create a new column.
+- Migrate existing data.
+- From next ESR, start using the new column.
+- Next ESR, drop the old column.
+
+For example, let’s say the next upcoming version is 8.4, and the next ESR is 8.6. So step 1 and 2, goes in 8.4. And in 8.7 onwards, we add the code to start using the new column, which will eventually be part of 8.12 (ESR after that). And then from 8.13 onwards, we can drop the column.
+
+The reasoning behind this is some customers will only upgrade from ESR to ESR. So we need to ensure backwards compatibility with the previous version.
+
+Following shows an example where we are adding a channel_count column to the status table. This is not exactly altering a column, but the idea remains the same, and you can extend this to fit your use-case.
+
+1. ALTER TABLE status ADD COLUMN channel_count integer;
+
+2. Our next objective is to migrate existing data. We do this in a 2-phase approach where we set up triggers to migrate all new data and in the background migrate old data in batches.
```

Comment:
How are we going to be sure the migration is completed with the trigger mechansim? I mean are we thinking of a mechanism to enable admisitrators to follow progress? Also can you elaborate how resilient is this mechanism?

Reply:
The trigger will only be executed for any new rows or updates. So there is no completion marker there. The completion marker is for the job, which will just add a new row in the Systems table when done. The reliability of that is handled by the jobs infrastructure, which we have been using since the start. The trigger resides in the DB itself, so no issues on that reliability wise.

Reply:
Oh okay, I missed the job part being handled by the application. Thanks for the clarification 👍